plugin/plugins/filesystem/nodes/ipfs.go

```diff
@@ -0,0 +1,199 @@
+
+	return id.Qid, req, metaClone, nil
+}
+
+func (id *IPFS) Walk(names []string) ([]p9.QID, p9.File, error) {
+	id.Logger.Debugf("ID Walk names %v", names)
+	id.Logger.Debugf("ID Walk myself: %s:%v", id.Path, id.Qid)
+
+	if shouldClone(names) {
+		id.Logger.Debugf("ID Walk cloned")
+		return []p9.QID{id.Qid}, id, nil
+	}
+
+	qids := make([]p9.QID, 0, len(names))
+
+	//TODO: [spec check] make sure we're not messing things up by doing this instead of mutating
+	// ^ Does internal library expect fid to mutate on success or does newfid clobber some external state anyway
+	walkedNode := &IPFS{} // operate on a copy
+	*walkedNode = *id
+
+	var err error
+	for _, name := range names {
+		//TODO: timerctx; don't want to hang forever
+		if walkedNode.Path, err = id.core.ResolvePath(id.Ctx, corepath.Join(walkedNode.Path, name)); err != nil {
+			return qids, nil, err
+		}
+
+		ipldNode, err := id.core.Dag().Get(id.Ctx, walkedNode.Path.Cid())
+		if err != nil {
+			return qids, nil, err
+		}
+
+		//TODO: this is too opague; we want core path => qid, dirent isn't necessary
+		dirEnt := &p9.Dirent{}
+		if err = ipldStat(dirEnt, ipldNode); err != nil {
+			return qids, nil, err
+		}
+
+		walkedNode.Qid = dirEnt.QID
+		//
+		qids = append(qids, walkedNode.Qid)
+	}
+
+	id.Logger.Debugf("ID Walk ret %v, %v", qids, walkedNode)
+	return qids, walkedNode, err
+}
+
+func (id *IPFS) Open(mode p9.OpenFlags) (p9.QID, uint32, error) {
+	id.Logger.Debugf("ID Open")
+	if id.meta.Mode.IsDir() {
+		c, err := id.core.Unixfs().Ls(id.Ctx, id.Path)
+		if err != nil {
+			return id.Qid, 0, err
+		}
+		id.directory = &directoryStream{
+			entryChan: c,
+		}
+	}
+
+	return id.Qid, ipfsBlockSize, nil
+}
+
+func (id *IPFS) Readdir(offset uint64, count uint32) ([]p9.Dirent, error) {
+	id.Logger.Debugf("ID Readdir")
+
+	if id.directory == nil {
+		return nil, fmt.Errorf("directory %q is not open for reading", id.Path.String())
+	}
+
+	if count == 0 {
+		return nil, nil
+	}
+
+	if offset < id.directory.cursor {
+		return nil, fmt.Errorf("read offset %d is behind current entry %d, seeking backwards in directory streams is not supported", offset, id.directory.cursor)
+	}
+
+	ents := make([]p9.Dirent, 0)
+
+out:
+	for len(ents) < int(count) {
+		select {
+		case entry, open := <-id.directory.entryChan:
+			if !open {
+				break out
+			}
+			if entry.Err != nil {
+				return nil, entry.Err
+			}
+
+			if offset <= id.directory.cursor {
+				nineEnt, err := coreEntTo9Ent(entry)
+				if err != nil {
+					return nil, err
+				}
+				nineEnt.Offset = id.directory.cursor
+				ents = append(ents, nineEnt)
+			}
+
+			id.directory.cursor++
+
+		case <-id.Ctx.Done():
+			return ents, id.Ctx.Err()
+		}
+	}
+
+	id.Logger.Debugf("ID Readdir returning [%d]%v\n", len(ents), ents)
+	return ents, nil
+}
+
+func (id *IPFS) ReadAt(p []byte, offset uint64) (int, error) {
+	id.Logger.Debugf("ID ReadAt")
+	const replaceMe = -1 //TODO [upstream/p9]: add constants for error 9P2000.L/Linux error values
```

Comment:
I don't know where I got it in my head that we needed the byte count to match a syscall errno when we return go errors.
This can be (0, err), but (-1, err) seems appropriate for a failed call.